sw/device/lib/base/BUILD

```diff
@@ -7,58 +7,107 @@ package(default_visibility = ["//visibility:public"])
 load("//rules:opentitan.bzl", "OPENTITAN_CPU")
 
 cc_library(
-    name = "base",
-    srcs = [
-        "bitfield.c",
-        "hardened.c",
-        "memory.c",
+    name = "stdasm",
+    hdrs = ["stdasm.h"],
+)
+
+cc_library(
+    name = "macros",
+    hdrs = ["macros.h"],
+)
+
+cc_library(
+    name = "bitfield",
+    srcs = ["bitfield.c"],
+    hdrs = ["bitfield.h"],
+)
+
+cc_library(
+    name = "memory",
+    srcs = ["memory.c"],
+    hdrs = ["memory.h"],
+
+    # This library defines memcpy(), so we can't have LLVM rewriting memcpy
+    # into a call to itself.
+    copts = ["-fno-builtin"],
+)
+
+cc_library(
+    name = "hardened",
+    srcs = ["hardened.c"],
+    hdrs = ["hardened.h"],
+    deps = [
+        ":memory",
+        ":stdasm",
     ],
+)
+
+cc_test(
+    name = "hardened_unittest",
+    srcs = ["hardened_unittest.cc"],
+    defines = [
+        "OT_OFF_TARGET_TEST",
+    ],
+    deps = [
+        ":hardened",
+        "@googletest//:gtest_main",
+    ],
+)
+
+cc_library(
+    name = "csr",
     hdrs = [
-        "bitfield.h",
         "csr.h",
         "csr_registers.h",
-        "hardened.h",
-        "macros.h",
-        "memory.h",
-        "mmio.h",
-        "multibits.h",
-        "multibits_asm.h",
-        "stdasm.h",
-    ],
-    copts = [
-        "-fno-builtin",
     ],
+    deps = [":stdasm"],
 )
 
 cc_library(
-    name = "abs_mmio",
-    srcs = [
-        "abs_mmio.c",
-    ],
+    name = "multibits",
     hdrs = [
-        "abs_mmio.h",
+        "multibits.h",
+        "multibits_asm.h",
     ],
-    target_compatible_with = [OPENTITAN_CPU],
+)
+
+cc_library(
+    name = "base",
+    srcs = [],
+    hdrs = ["mmio.h"],
+    deprecation = "Please use smaller lib/base targets instead of this one.",
     deps = [
-        ":base",
+        ":bitfield",
+        ":csr",
+        ":hardened",
+        ":macros",
+        ":memory",
+        ":multibits",
+        ":stdasm",
     ],
 )
 
 cc_library(
     name = "mmio",
-    srcs = [
-        "mmio.c",
-    ],
-    hdrs = [
-        "mmio.h",
-    ],
+    srcs = ["mmio.c"],
+    hdrs = ["mmio.h"],
     deps = [
-        ":base",
+        ":bitfield",
+        ":memory",
     ],
 )
 
+cc_library(
+    name = "abs_mmio",
+    srcs = ["abs_mmio.c"],
+    hdrs = ["abs_mmio.h"],
+    target_compatible_with = [OPENTITAN_CPU],
+    deps = [":mmio"],
+)
+
 # tests need to build src with a different set of flags. use this to do so.
 filegroup(
     name = "mmio_src",
     srcs = ["mmio.c"],
+    deprecation = "Please use :mmio instead, if possible.",
```

Comment:
If we depend on mmio how do we get the version built with MOCK_MMIO for unit tests and the version built without it for software built for OpenTitan?

Reply:
The MOCK_MMIO version should not depend on `mmio.c`, and provide those symbols itself. I'm going to see if I can just kill this target.